pkg/filter/filter.go

```diff
@@ -1,142 +1,96 @@
 package filter
 
 import (
-	"errors"
 	"fmt"
 	"os"
 	"reflect"
 	"strings"
 
+	"github.com/go-logr/logr"
 	"github.com/openshift-psap/special-resource-operator/pkg/clients"
 	"github.com/openshift-psap/special-resource-operator/pkg/color"
 	"github.com/openshift-psap/special-resource-operator/pkg/hash"
 	"github.com/openshift-psap/special-resource-operator/pkg/kernel"
 	"github.com/openshift-psap/special-resource-operator/pkg/lifecycle"
 	"github.com/openshift-psap/special-resource-operator/pkg/storage"
 	"github.com/openshift-psap/special-resource-operator/pkg/warn"
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 )
 
-const (
-	sroGVK = "SpecialResource"
-	owned  = "specialresource.openshift.io/owned"
-)
-
-var (
-	Mode string
-	log  = zap.New(zap.UseDevMode(true)).WithName(color.Print("filter", color.Purple))
-)
-
-func SetLabel(obj *unstructured.Unstructured) error {
-
-	var labels map[string]string
-
-	if labels = obj.GetLabels(); labels == nil {
-		labels = make(map[string]string)
-	}
-
-	labels[owned] = "true"
-	obj.SetLabels(labels)
-
-	return setSubResourceLabel(obj)
+type Filter interface {
+	GetPredicates() predicate.Predicate
+	GetMode() string
 }
 
-func setSubResourceLabel(obj *unstructured.Unstructured) error {
-
-	if obj.GetKind() == "DaemonSet" || obj.GetKind() == "Deployment" ||
-		obj.GetKind() == "StatefulSet" {
-
-		labels, found, err := unstructured.NestedMap(obj.Object, "spec", "template", "metadata", "labels")
-		if err != nil {
-			return err
-		}
-		if !found {
-			return errors.New("Labels not found")
-		}
-
-		labels[owned] = "true"
-		if err := unstructured.SetNestedMap(obj.Object, labels, "spec", "template", "metadata", "labels"); err != nil {
-			return err
-		}
+func NewFilter(kind, ownedLabel string) Filter {
+	return &filter{
+		log:        zap.New(zap.UseDevMode(true)).WithName(color.Print("filter", color.Purple)),
+		kind:       kind,
+		ownedLabel: ownedLabel,
 	}
+}
 
-	if obj.GetKind() == "BuildConfig" {
-		log.Info("TODO: how to set label ownership for Builds and related Pods")
```

Comment:
Do we want to keep this log?

It would require SetLabel and setSubResourceLabel to become a method to access creator's log

Reply:
Maybe we can make it a comment?

Reply:
it's there as a comment right now